detox/src/devices/DeviceRegistry.js

```diff
@@ -0,0 +1,75 @@
+const fs = require('fs');
+const plockfile = require('proper-lockfile');
+const _ = require('lodash');
+const retry = require('../utils/retry');
+const LOCK_FILE = './device.registry.state.lock';
+
+class DeviceRegistry {
+
+  constructor({getDeviceIdsByType, maxTestRunners = 1, createDevice}) {
+    this.getDeviceIdsByType = getDeviceIdsByType;
+    this.maxTestRunners = maxTestRunners;
+    this.createDevice = createDevice;
+    createEmptyLockFileIfNeeded();
+  }
+
+  async getDevice(deviceType) {
+    await retry(() => plockfile.lockSync(LOCK_FILE));
+    const deviceIds = await this.getDeviceIdsByType(deviceType);
+    await this._createDeviceIfNecessary({deviceIds, deviceType});
+
+    const unlockedDeviceId = getFirstUnlocked(deviceIds);
+    if (unlockedDeviceId) {
+      const lockedDevices = getLockedDevices();
+      lockedDevices.push(unlockedDeviceId);
+      writeLockedDevices(lockedDevices);
+      plockfile.unlockSync(LOCK_FILE);
+      return unlockedDeviceId;
+    }
+    plockfile.unlockSync(LOCK_FILE);
+    throw new Error(`Unable to find unlocked device ${deviceType}`);
+  }
+
+  static async freeDevice(deviceId) {
+    await retry(() => plockfile.lockSync(LOCK_FILE));
+    const lockedDevices = getLockedDevices();
+    _.remove(lockedDevices, lockedDeviceId => lockedDeviceId === deviceId);
+    writeLockedDevices(lockedDevices);
+    plockfile.unlockSync(LOCK_FILE);
+  }
+
+  async _createDeviceIfNecessary ({deviceIds, deviceType}) {
+    const numberOfDevicesNeededToCreate = Math.max(this.maxTestRunners - deviceIds.length, 0);
+    _.times(numberOfDevicesNeededToCreate, async () => await this.createDevice(deviceType));
```

Comment:
if `xcrun simctl` supports it, we may be able to create all devices simultaneously and use Promise.all to wait for all responses instead of doing them serially.

Reply:
I'm not sure. It's Xcode after all.